File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
import sys
input = sys.stdin.readline

n = int(input())
arr = []

def check(arr):
    cnt = 0
    for a in arr:
        if a[0]<=0:
            cnt+=1
    return cnt

def bt(v,arr):
    global res
    if v == n:
        res = max(res, check(arr))
        return

    if arr[v][0] <= 0:
        bt(v+1,arr)

    else:
        flag = True
        for i in range(n):
            if v != i and arr[i][0] >0:
                flag = False
                arr[v][0]-=arr[i][1]
                arr[i][0] -= arr[v][1]
                bt(v+1,arr)
                arr[v][0] += arr[i][1]
                arr[i][0] += arr[v][1]
        if flag:
            bt(n,arr)


for _ in range(n):
    a,b = map(int, input().split())
    arr.append([a,b])
res = 0
bt(0,arr)
print(res)

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
    static int answer;
    static int[][]arr;
    static int n;
    public static void main (String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        arr = new int[n][2];

        for(int i=0; i<n;i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            arr[i] = new int[]{Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())};
        }
        answer = 0;
        dfs(0,0);
        System.out.println(answer);
    }
    static void dfs(int L, int cnt){
        if(L== n||cnt==n-1){
            answer = Math.max(answer, cnt);
            return;
        }
        if(arr[L][0]<=0){
            dfs(L+1,cnt);
        }else{
        for(int i=0;i<n;i++){
            if(i==L)continue;
            if(arr[i][0]>0){
                arr[i][0]-=arr[L][1];
                arr[L][0] -=arr[i][1];
                dfs(L+1,cnt+(arr[i][0]<=0?1:0)+(arr[L][0]<=0?1:0));
                arr[i][0]+=arr[L][1];
                arr[L][0] +=arr[i][1];
            }
        }
    }


    }
}